### PR TITLE
Parses env cwl

### PR DESCRIPTION
**Pull request by danbills**
There was an issue w/ the declaration of the Requirement where one can either 

 * Specify the "class" of requirement with the Requirement itself
 * specify the class as a key to a map of requirements.

Decoders for `String => Requirement` type aliases were explicitly created to facilitate this.

### CODE REVIEW

File: src/main/scala/wdl4s/cwl/Implicits.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  implicit val dr = Decoder[DR]
+  implicit val scr = Decoder[SCR]
+  implicit val sfr = Decoder[SFR]
+  implicit val mifr = Decoder[MIFR]
+  implicit val sier = Decoder[SIER]
+
+  type V[A] = ValidatedNel[String,A]
+
+  def select[T](t: Target)(implicit selector: shapeless.ops.coproduct.Selector[Target,T]):ValidatedNel[String, T] =
+             t.select[T].toValidNel(s"Expecting a EnvVarRequirement but got $t instead.")
```
Comment (danbills) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: need to parameterize the actual expected requirement

File: src/main/scala/wdl4s/cwl/Implicits.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+  implicit val requirementArrayMapParser: Decoder[Array[Requirement]] =
```
Comment (danbills) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: document this

File: src/main/scala/wdl4s/cwl/TypeAliases.scala [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+trait TypeAliases {
+
+  type WorkflowStepInputId = String
+
+  type WorkflowStepInputSource = String :+: Array[String] :+: CNil
+
+  /**
+    * These are supposed to be valid ECMAScript Expressions.
+    * See http://www.commonwl.org/v1.0/Workflow.html#Expressions
+    */
```
Comment (danbills) [inline src/main/scala/wdl4s/cwl/TypeAliases.scala]: oh comments...

File: src/main/scala/wdl4s/cwl/Implicits.scala (hunk 2 above)
Comment (danbills) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: Requirement parser crazy enough to probably deserve it's own object?

File: src/main/scala/wdl4s/cwl/Implicits.scala [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            case ("SubworkflowFeatureRequirement", target) => select[SWFR](target).map(_("SubworkflowFeatureRequirement")).map(Coproduct[Requirement](_))
+            case ("ScatterFeatureRequirement", target) => select[SFR](target).map(_("ScatterFeatureRequirement")).map(Coproduct[Requirement](_))
+            case ("MultipleInputFeatureRequirement", target) => select[MIFR](target).map(_("MultipleInputFeatureRequirement")).map(Coproduct[Requirement](_))
+            case ("StepInputExpressionRequirement", target) => select[SIER](target).map(_("StepInputExpressionRequirement")).map(Coproduct[Requirement](_))
+            case (key,__) => invalidNel(s"key $key was not amongst possible values " +
+              "InlineJavascriptRequirement, SchemaDefRequirement, DockerRequirement, SoftwareRequirement, InitialWorkDirRequirement, EnvVarRequirement, ShellCommandRequirement, ResourceRequirement, SubworkflowFeatureRequirement, ScatterFeatureRequirement, MultipleInputFeatureRequirement, StepInputExpressionRequirement")
+          }.toEither.leftMap(_.toList.mkString(", ")).map(_.toArray)
+      }
```
Comment (danbills) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: This only supports map-based requirements.  Need to `or` w/ the default `Array[Requirement]` decoder

File: src/main/scala/wdl4s/cwl/Implicits.scala [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  implicit def iwdr = Decoder[IWDR]
+  implicit val dr = Decoder[DR]
+  implicit val scr = Decoder[SCR]
+  implicit val sfr = Decoder[SFR]
+  implicit val mifr = Decoder[MIFR]
+  implicit val sier = Decoder[SIER]
+
+  type V[A] = ValidatedNel[String,A]
+
+  def select[T](t: Target)(implicit selector: shapeless.ops.coproduct.Selector[Target,T]):ValidatedNel[String, T] =
```
Comment (mcovarr) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: can this return `ErrorOr[T]`?

File: src/main/scala/wdl4s/cwl/Implicits.scala [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+             t.select[T].toValidNel(s"Expecting a EnvVarRequirement but got $t instead.")
+
+  implicit val requirementArrayMapParser: Decoder[Array[Requirement]] =
+    Decoder[Map[String, Target]].
+      emap {
+        m =>
+          m.toList.traverse[V,Requirement] {
```
Comment (mcovarr) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: `_.toList...`?

File: src/main/scala/wdl4s/cwl/Implicits.scala [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            case ("SoftwareRequirement", target) => select[SR](target).map(_("SoftwareRequirement")).map(Coproduct[Requirement](_))
+            case ("InitialWorkDirRequirement", target) => select[IWDR](target).map(_("InitialWorkDirRequirement")).map(Coproduct[Requirement](_))
+            case ("ShellCommandRequirement", target) => select[SCR](target).map(_("ShellCommandRequirement")).map(Coproduct[Requirement](_))
+            case ("ResourceRequirement", target) => select[RR](target).map(_("ResourceRequirement")).map(Coproduct[Requirement](_))
+            case ("SubworkflowFeatureRequirement", target) => select[SWFR](target).map(_("SubworkflowFeatureRequirement")).map(Coproduct[Requirement](_))
+            case ("ScatterFeatureRequirement", target) => select[SFR](target).map(_("ScatterFeatureRequirement")).map(Coproduct[Requirement](_))
+            case ("MultipleInputFeatureRequirement", target) => select[MIFR](target).map(_("MultipleInputFeatureRequirement")).map(Coproduct[Requirement](_))
+            case ("StepInputExpressionRequirement", target) => select[SIER](target).map(_("StepInputExpressionRequirement")).map(Coproduct[Requirement](_))
```
Comment (mcovarr) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: it feels like there could be some DRYing here, at least the `.map(Coproduct...)` bit if not the preceding `map(_("string"))` as well

File: src/main/scala/wdl4s/cwl/model.scala [hunk 8]
```diff
@@ -36,9 +36,9 @@ case class InputParameter(
 }
 
 case class InputRecordSchema(
-  `type`: String Refined MatchesRegex[W.`"record"`.T],
+  `type`: W.`"record"`.T,
```
Comment (mcovarr) [inline src/main/scala/wdl4s/cwl/model.scala]: ✅

**Comment by mcovarr** [comment]
I think this should target cromwomification rather than develop.  This isn't needed on develop and would be more difficult to integrate into cromwomification when it is needed.

File: src/main/scala/wdl4s/cwl/Implicits.scala (hunk 2 above)
Comment (geoffjentry) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: This is certainly not our worst offender out there but I wouldn't be sad if you did it :)

**Comment by geoffjentry** [comment]
👍

File: src/main/scala/wdl4s/cwl/Implicits.scala [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            case (key,__) => invalidNel(s"key $key was not amongst possible values " +
+              "InlineJavascriptRequirement, SchemaDefRequirement, DockerRequirement, SoftwareRequirement, InitialWorkDirRequirement, EnvVarRequirement, ShellCommandRequirement, ResourceRequirement, SubworkflowFeatureRequirement, ScatterFeatureRequirement, MultipleInputFeatureRequirement, StepInputExpressionRequirement")
```
Comment (mcovarr) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: unfortunate the list of possible values needs to be rewritten here

File: src/main/scala/wdl4s/cwl/Implicits.scala (hunk 7 above)
Comment (danbills) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: Scout's honor I tried hard to DRY this up.  I'll declare something to shorten the `map`

What's hard to see here is the "string" argument is actually a different Witness type for each function in `Target`

File: src/main/scala/wdl4s/cwl/Implicits.scala (hunk 9 above)
Comment (danbills) [inline src/main/scala/wdl4s/cwl/Implicits.scala]: agree

**Comment by mcovarr** [comment]
:+1: you should merge this ahead of my subprojecting branch so you don't have to deal that rebase from hell 😄

**Comment by danbills** [comment]
per @mcovarr 's merge suggestion, I'm going to merge these changes but there are a few outstanding things to do.

I apologize, there is some cruft laying around, I'll be cleaning these up right away.

**Pull request merged**